db/snapshot/backup_task.cc

```diff
@@ -178,27 +202,32 @@ future<> backup_task_impl::process_snapshot_dir() {
     }
 }
 
-future<> backup_task_impl::uploads_worker() {
-    named_gate uploads(format("do_backup::uploads({})", _snapshot_dir));
+future<> backup_task_impl::worker::start_uploading() {
+    named_gate uploads(format("do_backup::uploads({})", _task._snapshot_dir));
 
     try {
         while (!_ex) {
             auto gh = uploads.hold();
-            auto units = co_await _sharded_worker.local().manager().dir_semaphore().get_units(1, _as);
+            auto units = co_await _manager.dir_semaphore().get_units(1, _as);
 
             // Pre-upload break point. For testing abort in actual s3 client usage.
             co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
 
-            auto name_opt = dequeue();
+            auto name_opt = co_await smp::submit_to(_task._backup_shard, [this] () {
+                return _task.dequeue();
+            });
             if (!name_opt) {
                 break;
             }
-            // okay to drop future since uploads is always closed before exiting the function
-            std::ignore = backup_file(std::move(*name_opt), upload_permit(std::move(gh), std::move(units)));
+            // okay to drop future since async_gate is always closed before stopping
+            std::ignore =
+                backup_file(std::move(*name_opt), upload_permit(std::move(gh), std::move(units))).then([this] {
+                    return _task.update_total_progress();
```

Comment:
This is horrible. Instead of bouncing to task shard once and updating the progress with two values in-place, you cause the task shard bounce back to _all_ other shards to collect the overall progress with map_reduce0

Reply:
It's better for the one who is asking for progress to bounce, since it's rare for progress to be queried.

Reply:
Agree, another option is to call map_reduce inside `get_progress()`. Just need to handle the case when workers are already dead (or still not alive)

Reply:
They can unlink themselves when they're done.